Translate Hooks FAQ

caal-15 opened the pull request:
Translated:

* Index
* First Section.

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+* **[Estrategia de Adopción](#adoption-strategy)**
+  * [¿Qué versiones de React incluyen Hooks?](#which-versions-of-react-include-hooks)
+  * [¿Necesito reescribir todos mis componentes que ya sean clases?](#do-i-need-to-rewrite-all-my-class-components)
+  * [¿Qué puedo hacer con los Hooks que no pueda hacer con las clases?](#what-can-i-do-with-hooks-that-i-couldnt-with-classes)
```
I suggest `¿Qué puedo hacer con Hooks que no pueda hacer con clases?`.

caal-15 replied:
Agreed, will change

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+  * [¿Qué tanto de mi conocimiento de React se mantiene relevante?](#how-much-of-my-react-knowledge-stays-relevant)
+  * [¿Debería usar Hooks, clases, o una mezcla de ambos?](#should-i-use-hooks-classes-or-a-mix-of-both)
+  * [¿Cubren los Hooks todos los casos de uso de las clases?](#do-hooks-cover-all-use-cases-for-classes)
+  * [¿Reemplazan los hooks a los render props y los Componente de Alto Orden (HOC)?](#do-hooks-replace-render-props-and-higher-order-components)
```
We're translating HOCs as `Componentes de Orden Superior`.

caal-15 replied:
Will Change

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+### ¿Cómo testear Componentes que usan Hooks? {#how-to-test-components-that-use-hooks}
 
-From React's point of view, a component using Hooks is just a regular component. If your testing solution doesn't rely on React internals, testing components with Hooks shouldn't be different from how you normally test components.
+Desde el punto de vista de React, un componente que use Hooks, sigue siendo un componente normal. Si las herramientas de testing que utilizas no depende de los mecanismos internos de React, testear los componentes que usen Hooks, no debería ser diferente de testear cualquier otro componente.
```
* no `dependen`.

caal-15 replied:
This was already changed :)

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+Desde el punto de vista de React, un componente que use Hooks, sigue siendo un componente normal. Si las herramientas de testing que utilizas no depende de los mecanismos internos de React, testear los componentes que usen Hooks, no debería ser diferente de testear cualquier otro componente.
 
-For example, let's say we have this counter component:
+Por ejemplo, asumamos que tenemos este componente de Conteo:
```
I suggest `componente contador`.

caal-15 replied:
I'm not that sure about this one, they convey the same meaning, though I think specifying it in terms of what the component does is better than specifying it in terms of what the component is

alejandronanez replied:
Agree with @caal-15, as they mean the same and it's proper Spanish. 👍

icarlossz commented on content/docs/hooks-faq.md:
````diff
@@ -171,44 +173,44 @@ it('can render and update a counter', () => {
 ```
 
-The calls to `act()` will also flush the effects inside of them.
+Las llamadas a `act()` también resolverán los effectos adentro de ellas.
````
* `efectos`

caal-15 replied:
Goddamn Spanglish xD, wil change :)

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -171,44 +173,44 @@ it('can render and update a counter', () => {
+Las llamadas a `act()` también resolverán los effectos adentro de ellas.
 
-If you need to test a custom Hook, you can do so by creating a component in your test, and using your Hook from it. Then you can test the component you wrote.
+Si necesitas testear un Hook personalizado, puedes hacerlo creando un componente en tu test, y usando tu Hook desde el mismo. Luego puedes testear el componente que escribiste.
```
* `él mismo`.

icarlossz replied:
👀

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -171,44 +173,44 @@ it('can render and update a counter', () => {
+Si necesitas testear un Hook personalizado, puedes hacerlo creando un componente en tu test, y usando tu Hook desde el mismo. Luego puedes testear el componente que escribiste.
 
-To reduce the boilerplate, we recommend using [`react-testing-library`](https://git.io/react-testing-library) which is designed to encourage writing tests that use your components as the end users do.
+Para reducir el boilerplate, recomendamos usar [`react-testing-library`](https://git.io/react-testing-library) que está diseñada para promover tests que utilicesn tus componentes como lo harían los usuarios finales.
```
* `utilicen`.

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -171,44 +173,44 @@ it('can render and update a counter', () => {
+Proveemos un [plugin de ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks) que hace cumplir las [reglas de los Hooks](/docs/hooks-rules.html) para evitar bugs. Asume que cualquier función cuyo nombre empiece con "`use`", seguido de una letra mayúscula es un Hook. Reconocemos que esta heurística no es perfecta, y podría haber algunos falsos positivos, pero sin una convención que cubr ...
 
-In particular, the rule enforces that:
+En particular, la regla hace cumplir que:
```
I'd put `la regla hace cumplir` or `la regla fuerza que`.

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -171,44 +173,44 @@ it('can render and update a counter', () => {
+### ¿Existe algo similar a las variables de instancia? {#is-there-something-like-instance-variables}
 
-Yes! The [`useRef()`](/docs/hooks-reference.html#useref) Hook isn't just for DOM refs. The "ref" object is a generic container whose `current` property is mutable and can hold any value, similar to an instance property on a class.
+Si!, el Hook [`useRef()`](/docs/hooks-reference.html#useref) no es solo para referencias al DOM. El objeto "ref" es un contenedor genérico cuya propiedad `current` es mutable y puede contener cualquier valor, similar a una variable de instancia en una clase.
```
In Spanish the exclamation signs should also be placed at the beginning so should be `¡Si!`.

icarlossz commented on content/docs/hooks-faq.md:
````diff
@@ -238,11 +240,11 @@ If we just wanted to set an interval, we wouldn't need the ref (`id` could be lo
 ```
 
-Conceptually, you can think of refs as similar to instance variables in a class. Unless you're doing [lazy initialization](#how-to-create-expensive-objects-lazily), avoid setting refs during rendering -- this can lead to surprising behavior. Instead, typically you want to modify refs in event handlers and effects.
+Conceptualmente, puedes pensar en los refs como símiles a las variables de instancia en una clase. A menos que estés utilizando inicialización perezosa ([lazy initialization](#how-to-create-expensive-objects-lazily)), evita setear referencias durante el renderizado -- esto podría llevar a comportamiento inesperado. En cambio, generalmente querrás modificar las referencias en manejadores de evento ...
````
I suggest `puedes pensar en las referencias como algo similar a..`

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -267,11 +269,11 @@ Now let's say we want to write some logic that changes `left` and `top` when the
+Si extrañas esta mezcla automática, puedes escribir un Hook personalizado `useLegacyState` que mezcle las actualizaciones al objeto de estado. Sin embargo, **Recomendamos dividir el estado en múltiples variables de estado, basado en los valores que tienden a cambiar juntos**.
 
-For example, we could split our component state into `position` and `size` objects, and always replace the `position` with no need for merging:
+Por ejemplo, podríamos dividir el estado de nuestor componente en objetos `position` y `size`, y siempre reemplazar `position` sin la necesidad de mezclar.
```
* `nuestro`.

alejandronanez replied:
```suggestion
Por ejemplo, podríamos dividir el estado de nuestro componente en objetos `position` y `size`, y siempre reemplazar `position` sin la necesidad de mezclar.
```

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -303,17 +305,17 @@ function useWindowPosition() {
+Ambas aproximaciones, poner todo el estado en un solo llamdo a `useState`, y usar un llamado a `useState` por cada campo, pueden funcionar. Los Componentes suelen ser más legibles cuando encuentras un balance entre ambos extremos, y agrupas partes del estado relacionadas en unas cuantas variables de estado independientes. Si la lógica del estado se vuelve muy compleja, recomendamos [manejarla con ...
 
-### Can I run an effect only on updates? {#can-i-run-an-effect-only-on-updates}
+### ¿Puedo correr un efecto solo en cuando hayan actualizaciones? {#can-i-run-an-effect-only-on-updates}
```
I suggest `¿Puedo correr un efecto solo cuando hayan actualizaciones?`

caal-15 replied:
Yes this was definitley something I missed, thanks!

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -303,17 +305,17 @@ function useWindowPosition() {
+Este es un caso de uso poco común. Si lo necesitas, puedes usar [una referencia mutable](#is-there-something-like-instance-variables) para guardar manualmente una bandera booleana que corresponde a si es el primer renderizado, o renderizados subsecuentes, luego puedes verificar la bandera en tu efecto. (Si te encuentras haciendo esto regularmente podrías crear un Hook Personalizado).
 
-### How to get the previous props or state? {#how-to-get-the-previous-props-or-state}
+### ¿Cómo obtengo los props o el estado previo? {#how-to-get-the-previous-props-or-state}
```
Should be `las props`.

caal-15 replied:
Will change

icarlossz commented on content/docs/hooks-faq.md:
````diff
@@ -347,7 +349,7 @@ function usePrevious(value) {
 ```
 
-Note how this would work for props, state, or any other calculated value.
+Nota como esto podría funcionar para props, estado, o cualquie rotro valor calculado.
````
* `cualquier otro`.

caal-15 replied:
Will Change!

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -358,15 +360,15 @@ function Counter() {
+Es posible que en el futuro React provea un `usePrevious` Hook por defecto, ya que es un caso de uso relativamente común.
 
-See also [the recommended pattern for derived state](#how-do-i-implement-getderivedstatefromprops).
+Mira también [el patrón recomendado para estado derivado](#how-do-i-implement-getderivedstatefromprops).
```
* `para un estado derivado`

caal-15 replied:
Will Change!

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -383,13 +385,13 @@ function ScrollView({row}) {
+Los Hooks `useState` y `useReducer` [evitan las actualizaciones](/docs/hooks-reference.html#bailing-out-of-a-state-update) si el siguiente valor es igual al anterior. Mutar el estado y llamar a `setState` no causarán un re-renderizado.
 
-Normally, you shouldn't mutate local state in React. However, as an escape hatch, you can use an incrementing counter to force a re-render even if the state has not changed:
+Usualmente, nodeberías mutar el estado local en React. Sin embargo, como una salida de emergencia, puedes usar un contador incremental para forzar un re-renderizado incluso si el estado no ha cambiado:
```
* `no deberías`

caal-15 replied:
Will Change!

caal-15 commented on content/docs/hooks-faq.md:
```diff
@@ -5,9 +5,11 @@ permalink: docs/hooks-faq.html
 
-*Hooks* are a new addition in React 16.8. They let you use state and other React features without writing a class.
+Los *Hooks* son una adición nueva en React 16.8. Te permiten user el estado y
+otras caracterísitcas de React sin la necesidad de escribir una clase.
```
Thanks, will correct!

caal-15 commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+  * [¿Reemplazan los hooks a los render props y los Componente de Alto Orden (HOC)?](#do-hooks-replace-render-props-and-higher-order-components)
+  * [¿Qué significan los Hooks para APIs populares como el connect de Redux, o React Router?](#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router)
+  * [¿Funcionan los Hooks con tipado estático?](#do-hooks-work-with-static-typing)
+  * [¿Cómo testear Componentes que usan Hooks?](#how-to-test-components-that-use-hooks)
```
I thought about it too, probably better to change it

caal-15 commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+  * [¿Qué significan los Hooks para APIs populares como el connect de Redux, o React Router?](#what-do-hooks-mean-for-popular-apis-like-redux-connect-and-react-router)
+  * [¿Funcionan los Hooks con tipado estático?](#do-hooks-work-with-static-typing)
+  * [¿Cómo testear Componentes que usan Hooks?](#how-to-test-components-that-use-hooks)
+  * [¿Qué hacen cumplir las reglas de lint?](#what-exactly-do-the-lint-rules-enforce)
```
Ummm I'm not sure about this one "fuerzan" could be given a different meaning than what this intends... anyone else cares to chime in?

caal-15 commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+  * [¿Cómo corresponden los métodos del ciclo de vida a los Hooks?](#how-do-lifecycle-methods-correspond-to-hooks)
+  * [¿Existe algo similar a las variables de instancia?](#is-there-something-like-instance-variables)
+  * [¿Debería usar una o muchas variables de estado?](#should-i-use-one-or-many-state-variables)
+  * [¿Puedo correr un efecto solo en cuando hayan actualizaciones?](#can-i-run-an-effect-only-on-updates)
```
Agreed, will change

caal-15 commented:
> * `Componentes`

Can you point me at the line where you find a mistake?

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+  * [¿Qué tanto de mi conocimiento de React se mantiene relevante?](#how-much-of-my-react-knowledge-stays-relevant)
+  * [¿Debería usar Hooks, clases, o una mezcla de ambos?](#should-i-use-hooks-classes-or-a-mix-of-both)
+  * [¿Cubren los Hooks todos los casos de uso de las clases?](#do-hooks-cover-all-use-cases-for-classes)
+  * [¿Reemplazan los hooks a los render props y los Componente de Orden Superior (HOC)?](#do-hooks-replace-render-props-and-higher-order-components)
```
* `Componentes`.

caal-15 replied:
Thanks, will change!

icarlossz replied:
don't forget the hidden comments above

alejandronanez commented on content/docs/hooks-faq.md:
```diff
@@ -358,15 +360,15 @@ function Counter() {
+Mira también [el patrón recomendado para un estado derivado](#how-do-i-implement-getderivedstatefromprops).
 
-### How do I implement `getDerivedStateFromProps`?
+### ¿Cómo implemento getDerivedStateFromProps??
```
```suggestion
### ¿Cómo implemento getDerivedStateFromProps?
```

caal-15 replied:
Will Change

sergiodxa commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+### ¿Debería usar Hooks, clases, o una mezcla de ambos? {#should-i-use-hooks-classes-or-a-mix-of-both}
 
-When you're ready, we'd encourage you to start trying Hooks in new components you write. Make sure everyone on your team is on board with using them and familiar with this documentation. We don't recommend rewriting your existing classes to Hooks unless you planned to rewrite them anyway (e.g. to fix bugs).
+Cuando estés listo, te recomendamos empezar a usar Hooks en los nuevos componentes que escribas. Asegúrate que todo tu equipo esté de acuerdo en usarlos, y que estén familiarizados con esta documentación. No recomendamos reescribir tus classes exitentes a menos de que hayas planeado reescribirlas de cualquier manera (por ejemplo para arreglar bugs).
```
Also `classes` should be `clases`

caal-15 replied:
Will change!

alejandronanez commented:
Hey @caal-15, heads up, there are some merge conflicts that you have to resolve before merging!

caal-15 commented on content/docs/hooks-faq.md:
```diff
@@ -5,9 +5,11 @@ permalink: docs/hooks-faq.html
 ---
 
-*Hooks* are a new addition in React 16.8. They let you use state and other React features without writing a class.
+Los *Hooks* son una adición nueva en React 16.8. Te permiten user el estado y
```
Will do!

caal-15 commented on content/docs/hooks-faq.md:
```diff
@@ -18,102 +20,102 @@ This page answers some of the frequently asked questions about [Hooks](/docs/hoo
+### ¿Reemplazan los hooks a los render props y los Componentes de Orden Superior (HOC)? {#do-hooks-replace-render-props-and-higher-order-components}
 
-Often, render props and higher-order components render only a single child. We think Hooks are a simpler way to serve this use case. There is still a place for both patterns (for example, a virtual scroller component might have a `renderItem` prop, or a visual container component might have its own DOM structure). But in most cases, Hooks will be sufficient and can help reduce nesting in your tre ...
+En muchas ocasiones, render props y los componentes de orden superior, renderizan un sólo hijo. Pensamos que los Hooks son una forma más sencilla de soportar este caso de uso. Aún hay lugar para ambos patrones (por ejemplo, un scroller virtual podría tener un prop `renderItem`, o un conmponente que sea un contenedor visual podría tener su propia estructura de DOM). Pero en la mayoría de los casos ...
```
Will Change!

caal-15 commented:
Hey everyone!, I have fixed the merge conflicts and added @alejandronanez's suggestions, I can add the las section today so we have the whole thing ready for a final review :D

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -411,48 +411,47 @@ A pesar de que no deberías necesitar esto muy seguido, podrías exponer algunos
+Recuerda que la función que se pasa a `useMemo` corre durante el renderizado. No hagas nada allí que no harías durante el renderizado. Por ejemplo, los efectos secundarios deberían estar en `useEffect`, no en `useMemo`.
 
-**You may rely on `useMemo` as a performance optimization, not as a semantic guarantee.** In the future, React may choose to "forget" some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without `useMemo` — and then add it to optimize performance. (For rare cases when a value must *never* be recom ...
+**Puedes depender de `useMemo` como una mejora de desempeño, pero no como una garantía semántica.** In el futuro, React podría escoger "olvidar" algunos valores previamente memorizados y recalcularlos en el siguiente renderizado, por ejemplo para liberar memoria para los components que no se ven en pantalla. Escribe to código de manera que pueda funcionar sin `useMemo` — y luego añádelo para mejo ...
```
* `En el futuro`

caal-15 replied:
Will Change!

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -463,51 +462,51 @@ function Parent({ a, b }) {
+Ten en cuenta que este método no funcionará en un ciclo porque las llamadas a Hooks [no pueden](/docs/hooks-rules.html) ser puestas dentro de ciclos. Pero puedes extraer un componente separado para el item de la lista, y llamar `useMemo` allí.
 
-### How to create expensive objects lazily? {#how-to-create-expensive-objects-lazily}
+### ¿Cómo crear objetos costosos de manera perezosa (lazy)? {#how-to-create-expensive-objects-lazily}
```
Idk how will be the right way to translate this but it looks pretty rare. I'd put `¿Cómo crear objetos pesados de manera perezosa/floja (lazy)?`

carburo replied:
El término preferido en el glosario para «*lazy initialization*» es «inicialización diferida». Por tanto, por consistencia, probablemente lo mejor sería traducirlo de la siguiente forma: ¿Cómo crear objetos costosos de manera diferida?

icarlossz replied:
Yeah, much better. // Si, está mucho mejor.

alejandronanez replied:
```suggestion
### ¿Cómo crear objetos costosos de manera diferida (lazy)? {#how-to-create-expensive-objects-lazily}
```

caal-15 replied:
Will change, this also applies in some other places, Will change those also for consistency

icarlossz commented on content/docs/hooks-faq.md:
````diff
@@ -463,51 +462,51 @@ function Parent({ a, b }) {
 ```js
 function Table(props) {
-  // ✅ createRows() is only called once
+  // ✅ createRows() solo se llama una vez.
````
I suggest `es llamado una sola vez`.

alejandronanez replied:
I think this is fine in both ways, either `es llamado una sola vez` or `solo se llama una vez`. @caal-15, feel free to change it or leave it as is - I don't have strong opinions about this change! :)

caal-15 replied:
Will leave it as it is :)

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -518,50 +517,50 @@ function Image(props) {
+Adicionalmente, considera que el diseño de los Hooks es más eficiente en un par de sentidos:
 
-* Hooks avoid a lot of the overhead that classes require, like the cost of creating class instances and binding event handlers in the constructor.
+* Evitan gran parte del overhead (trabajo extra) que las clases requieren, como el costo de crear instancias de clase y ligar (bind) los manejadores de eventos en el constructor.
```
I'd put `Evitan gran parte de la sobrecarga que las clases requieren`

alejandronanez replied:
```suggestion
* Evitan gran parte de la complejidad (trabajo extra) que las clases requieren, como el costo de crear instancias de clase y ligar (bind) los manejadores de eventos en el constructor.
```

caal-15 replied:
Will change :)

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -518,50 +517,50 @@ function Image(props) {
+* Evitan gran parte del overhead (trabajo extra) que las clases requieren, como el costo de crear instancias de clase y ligar (bind) los manejadores de eventos en el constructor.
 
-* **Idiomatic code using Hooks doesn't need the deep component tree nesting** that is prevalent in codebases that use higher-order components, render props, and context. With smaller component trees, React has less work to do.
+* **El código idiómatico usando Hooks no requiere el anidado profundo de componentes** que es prevalente en bases de código que utilizan componente de orden superior, render props, y contexto. Con árboles de componentes más pequeños, React tiene menos trabajo que realizar.
```
* `componentes` de orden superior.

caal-15 replied:
Will change.

icarlossz commented on content/docs/hooks-faq.md:
```diff
@@ -518,50 +517,50 @@ function Image(props) {
+* **El código idiómatico usando Hooks no requiere el anidado profundo de componentes** que es prevalente en bases de código que utilizan componente de orden superior, render props, y contexto. Con árboles de componentes más pequeños, React tiene menos trabajo que realizar.
 
-Traditionally, performance concerns around inline functions in React have been related to how passing new callbacks on each render breaks `shouldComponentUpdate` optimizations in child components. Hooks approach this problem from three sides.
+Tradicionalmente, las preocupaciones de desempeño alrededor de funciones inline en React han estado relacionadas con como al pasar nuevos callbacks en cada renderizado rompe optimizaciones con `shouldComponentUpdate` en los componentes hijo. Los Hooks pueden resolver este problema desde tres ángulos diferentes.
```
* componentes `hijos`

caal-15 replied:
Will change :)

alejandronanez commented on content/docs/hooks-faq.md:
```diff
@@ -464,51 +462,51 @@ function Parent({ a, b }) {
   const ref = useRef(null);
 
-  // ✅ IntersectionObserver is created lazily once
+  // ✅ IntersectionObserver se crea perezosamente una vez.
```
```suggestion
  // ✅ IntersectionObserver se crea diferidamente una vez.
```

caal-15 replied:
I have already changed 'perezoso' for 'diferido' in general :)
